ICE: `expected usize, got Const`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code
` rustc --edition=2021 file.rs`
```Rust
#![feature(generic_const_exprs)]
#![crate_type="lib"]

pub async fn something(path: &[[usize; N_ISLANDS]; N_ISLANDS]) -> usize {
    async {
        match path {
            [] => 0,

            _ => 1,
        }
    }
    .await
}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.72.0-nightly (df77afbca 2023-06-12)
binary: rustc
commit-hash: df77afbcaf3365a32066a8ca4a00ae6fc9a69647
commit-date: 2023-06-12
host: x86_64-unknown-linux-gnu
release: 1.72.0-nightly
LLVM version: 16.0.5
```

### Error output

```
error[E0425]: cannot find value `N_ISLANDS` in this scope
 --> treereduce.out:4:40
  |
4 | pub async fn something(path: &[[usize; N_ISLANDS]; N_ISLANDS]) -> usize {
  |                                        ^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `N_ISLANDS` in this scope
 --> treereduce.out:4:52
  |
4 | pub async fn something(path: &[[usize; N_ISLANDS]; N_ISLANDS]) -> usize {
  |                                                    ^^^^^^^^^ not found in this scope

warning: the feature `generic_const_exprs` is incomplete and may not be safe to use and/or cause compiler crashes
 --> treereduce.out:1:12
  |
1 | #![feature(generic_const_exprs)]
  |            ^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #76560 <https://github.com/rust-lang/rust/issues/76560> for more information
  = note: `#[warn(incomplete_features)]` on by default
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/compiler/rustc_middle/src/ty/consts.rs:238:32: expected usize, got Const { ty: usize, kind: Unevaluated([], DefId(0:5 ~ treereduce[d695]::something::{constant#1})) }

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/compiler/rustc_errors/src/lib.rs:1650:9
stack backtrace:
   0:     0x7f6ac3d93c41 - std::backtrace_rs::backtrace::libunwind::trace::hd830ec006d66d7bd
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f6ac3d93c41 - std::backtrace_rs::backtrace::trace_unsynchronized::ha48a9544f0edee03
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f6ac3d93c41 - std::sys_common::backtrace::_print_fmt::had4ddaa258f461e3
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f6ac3d93c41 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::he768f658e406b581
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f6ac3df45ff - core::fmt::rt::Argument::fmt::h10ce9a0440e62ef5
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/core/src/fmt/rt.rs:138:9
   5:     0x7f6ac3df45ff - core::fmt::write::h381a5b53ed990481
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/core/src/fmt/mod.rs:1094:21
   6:     0x7f6ac3d86e41 - std::io::Write::write_fmt::h6f7ffab47f774bac
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/std/src/io/mod.rs:1713:15
   7:     0x7f6ac3d93a55 - std::sys_common::backtrace::_print::h6b3b514a9c0bb2a1
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f6ac3d93a55 - std::sys_common::backtrace::print::h4d04f57fcaaadbf7
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f6ac3d96737 - std::panicking::default_hook::{{closure}}::h82f2d8896eabc353
  10:     0x7f6ac3d96524 - std::panicking::default_hook::h1d70e3392ff8715d
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/std/src/panicking.rs:288:9
  11:     0x7f6ac2beb80b - rustc_driver_impl[bc9e93faacc37431]::install_ice_hook::{closure#0}
  12:     0x7f6ac3d96e7d - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hdeb68781aec8c01b
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/alloc/src/boxed.rs:1999:9
  13:     0x7f6ac3d96e7d - std::panicking::rust_panic_with_hook::h411bd05f08be05b6
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/std/src/panicking.rs:709:13
  14:     0x7f6ac3101321 - std[74cc9e3c59aa80ae]::panicking::begin_panic::<rustc_errors[c5ec46bf3b0ea9c2]::ExplicitBug>::{closure#0}
  15:     0x7f6ac30fcdf6 - std[74cc9e3c59aa80ae]::sys_common::backtrace::__rust_end_short_backtrace::<std[74cc9e3c59aa80ae]::panicking::begin_panic<rustc_errors[c5ec46bf3b0ea9c2]::ExplicitBug>::{closure#0}, !>
  16:     0x7f6ac31781c6 - std[74cc9e3c59aa80ae]::panicking::begin_panic::<rustc_errors[c5ec46bf3b0ea9c2]::ExplicitBug>
  17:     0x7f6ac30f75c4 - <rustc_errors[c5ec46bf3b0ea9c2]::HandlerInner>::bug::<alloc[87cbe4d25bb98c8e]::string::String>
  18:     0x7f6ac30f7366 - <rustc_errors[c5ec46bf3b0ea9c2]::Handler>::bug::<alloc[87cbe4d25bb98c8e]::string::String>
  19:     0x7f6ac31865b5 - rustc_middle[111bd832a8dd3e9b]::util::bug::opt_span_bug_fmt::<rustc_span[bc106c13571faf85]::span_encoding::Span>::{closure#0}
  20:     0x7f6ac31831ca - rustc_middle[111bd832a8dd3e9b]::ty::context::tls::with_opt::<rustc_middle[111bd832a8dd3e9b]::util::bug::opt_span_bug_fmt<rustc_span[bc106c13571faf85]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7f6ac318319a - rustc_middle[111bd832a8dd3e9b]::ty::context::tls::with_context_opt::<rustc_middle[111bd832a8dd3e9b]::ty::context::tls::with_opt<rustc_middle[111bd832a8dd3e9b]::util::bug::opt_span_bug_fmt<rustc_span[bc106c13571faf85]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7f6ac0e6534d - rustc_middle[111bd832a8dd3e9b]::util::bug::bug_fmt
  23:     0x7f6ac1aee5b8 - <rustc_mir_build[2d2c8624bc56a64a]::build::Builder>::prefix_slice_suffix
  24:     0x7f6ac1ae8e5a - <rustc_mir_build[2d2c8624bc56a64a]::build::Builder>::match_candidates
  25:     0x7f6ac0f36fc7 - <rustc_mir_build[2d2c8624bc56a64a]::build::Builder>::lower_match_tree
  26:     0x7f6ac0f2391c - <rustc_mir_build[2d2c8624bc56a64a]::build::Builder>::expr_into_dest
  27:     0x7f6ac0f218ab - <rustc_mir_build[2d2c8624bc56a64a]::build::Builder>::expr_into_dest
  28:     0x7f6ac0f3fba8 - <rustc_mir_build[2d2c8624bc56a64a]::build::Builder>::ast_block_stmts
  29:     0x7f6ac0f35863 - <rustc_mir_build[2d2c8624bc56a64a]::build::Builder>::ast_block
  30:     0x7f6ac0f2296f - <rustc_mir_build[2d2c8624bc56a64a]::build::Builder>::expr_into_dest
  31:     0x7f6ac0f218ab - <rustc_mir_build[2d2c8624bc56a64a]::build::Builder>::expr_into_dest
  32:     0x7f6ac0f218ab - <rustc_mir_build[2d2c8624bc56a64a]::build::Builder>::expr_into_dest
  33:     0x7f6ac1a729fe - rustc_mir_build[2d2c8624bc56a64a]::build::mir_built
  34:     0x7f6ac0c0d44c - rustc_query_impl[3b08f3ae92a78a66]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_built::dynamic_query::{closure#2}::{closure#0}, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>
  35:     0x7f6ac0c0d42e - <rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_built::dynamic_query::{closure#2} as core[7eaf5f0da9577636]::ops::function::FnOnce<(rustc_middle[111bd832a8dd3e9b]::ty::context::TyCtxt, rustc_span[bc106c13571faf85]::def_id::LocalDefId)>>::call_once
  36:     0x7f6ac0cf4b6d - rustc_query_system[1b62b8d1dafbbc4d]::query::plumbing::try_execute_query::<rustc_query_impl[3b08f3ae92a78a66]::DynamicConfig<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::VecCache<rustc_span[bc106c13571faf85]::def_id::LocalDefId, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3b08f3ae92a78a66]::plumbing::QueryCtxt, false>
  37:     0x7f6ac25676d1 - rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_built::get_query_non_incr::__rust_end_short_backtrace
  38:     0x7f6ac1b18cc4 - rustc_mir_transform[3aed451d879c1d8b]::check_unsafety::unsafety_check_result
  39:     0x7f6ac0c0d48c - rustc_query_impl[3b08f3ae92a78a66]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3b08f3ae92a78a66]::query_impl::unsafety_check_result::dynamic_query::{closure#2}::{closure#0}, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>
  40:     0x7f6ac0c0d46e - <rustc_query_impl[3b08f3ae92a78a66]::query_impl::unsafety_check_result::dynamic_query::{closure#2} as core[7eaf5f0da9577636]::ops::function::FnOnce<(rustc_middle[111bd832a8dd3e9b]::ty::context::TyCtxt, rustc_span[bc106c13571faf85]::def_id::LocalDefId)>>::call_once
  41:     0x7f6ac0cf4b6d - rustc_query_system[1b62b8d1dafbbc4d]::query::plumbing::try_execute_query::<rustc_query_impl[3b08f3ae92a78a66]::DynamicConfig<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::VecCache<rustc_span[bc106c13571faf85]::def_id::LocalDefId, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3b08f3ae92a78a66]::plumbing::QueryCtxt, false>
  42:     0x7f6ac2569581 - rustc_query_impl[3b08f3ae92a78a66]::query_impl::unsafety_check_result::get_query_non_incr::__rust_end_short_backtrace
  43:     0x7f6ac1b1bdcc - rustc_mir_transform[3aed451d879c1d8b]::check_unsafety::unsafety_check_result
  44:     0x7f6ac0c0d48c - rustc_query_impl[3b08f3ae92a78a66]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3b08f3ae92a78a66]::query_impl::unsafety_check_result::dynamic_query::{closure#2}::{closure#0}, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>
  45:     0x7f6ac0c0d46e - <rustc_query_impl[3b08f3ae92a78a66]::query_impl::unsafety_check_result::dynamic_query::{closure#2} as core[7eaf5f0da9577636]::ops::function::FnOnce<(rustc_middle[111bd832a8dd3e9b]::ty::context::TyCtxt, rustc_span[bc106c13571faf85]::def_id::LocalDefId)>>::call_once
  46:     0x7f6ac0cf4b6d - rustc_query_system[1b62b8d1dafbbc4d]::query::plumbing::try_execute_query::<rustc_query_impl[3b08f3ae92a78a66]::DynamicConfig<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::VecCache<rustc_span[bc106c13571faf85]::def_id::LocalDefId, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3b08f3ae92a78a66]::plumbing::QueryCtxt, false>
  47:     0x7f6ac2569581 - rustc_query_impl[3b08f3ae92a78a66]::query_impl::unsafety_check_result::get_query_non_incr::__rust_end_short_backtrace
  48:     0x7f6ac1b1bdcc - rustc_mir_transform[3aed451d879c1d8b]::check_unsafety::unsafety_check_result
  49:     0x7f6ac0c0d48c - rustc_query_impl[3b08f3ae92a78a66]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3b08f3ae92a78a66]::query_impl::unsafety_check_result::dynamic_query::{closure#2}::{closure#0}, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>
  50:     0x7f6ac0c0d46e - <rustc_query_impl[3b08f3ae92a78a66]::query_impl::unsafety_check_result::dynamic_query::{closure#2} as core[7eaf5f0da9577636]::ops::function::FnOnce<(rustc_middle[111bd832a8dd3e9b]::ty::context::TyCtxt, rustc_span[bc106c13571faf85]::def_id::LocalDefId)>>::call_once
  51:     0x7f6ac0cf4b6d - rustc_query_system[1b62b8d1dafbbc4d]::query::plumbing::try_execute_query::<rustc_query_impl[3b08f3ae92a78a66]::DynamicConfig<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::VecCache<rustc_span[bc106c13571faf85]::def_id::LocalDefId, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3b08f3ae92a78a66]::plumbing::QueryCtxt, false>
  52:     0x7f6ac2569581 - rustc_query_impl[3b08f3ae92a78a66]::query_impl::unsafety_check_result::get_query_non_incr::__rust_end_short_backtrace
  53:     0x7f6ac1f93504 - rustc_mir_transform[3aed451d879c1d8b]::mir_const
  54:     0x7f6ac0c22f6e - rustc_query_impl[3b08f3ae92a78a66]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_const::dynamic_query::{closure#2}::{closure#0}, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>
  55:     0x7f6ac0c22f3e - <rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_const::dynamic_query::{closure#2} as core[7eaf5f0da9577636]::ops::function::FnOnce<(rustc_middle[111bd832a8dd3e9b]::ty::context::TyCtxt, rustc_span[bc106c13571faf85]::def_id::LocalDefId)>>::call_once
  56:     0x7f6ac0cf4b6d - rustc_query_system[1b62b8d1dafbbc4d]::query::plumbing::try_execute_query::<rustc_query_impl[3b08f3ae92a78a66]::DynamicConfig<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::VecCache<rustc_span[bc106c13571faf85]::def_id::LocalDefId, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3b08f3ae92a78a66]::plumbing::QueryCtxt, false>
  57:     0x7f6ac25677d1 - rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_const::get_query_non_incr::__rust_end_short_backtrace
  58:     0x7f6ac1a3dfac - rustc_mir_transform[3aed451d879c1d8b]::mir_promoted
  59:     0x7f6ac0bf70cf - rustc_query_impl[3b08f3ae92a78a66]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_promoted::dynamic_query::{closure#2}::{closure#0}, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 16usize]>>
  60:     0x7f6ac146a67a - rustc_query_system[1b62b8d1dafbbc4d]::query::plumbing::try_execute_query::<rustc_query_impl[3b08f3ae92a78a66]::DynamicConfig<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::VecCache<rustc_span[bc106c13571faf85]::def_id::LocalDefId, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[3b08f3ae92a78a66]::plumbing::QueryCtxt, false>
  61:     0x7f6ac2567c17 - rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_promoted::get_query_non_incr::__rust_end_short_backtrace
  62:     0x7f6ac1b24bdc - rustc_borrowck[5f856fd74d412c64]::mir_borrowck
  63:     0x7f6ac0c22f0e - rustc_query_impl[3b08f3ae92a78a66]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>
  64:     0x7f6ac0c22ede - <rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_borrowck::dynamic_query::{closure#2} as core[7eaf5f0da9577636]::ops::function::FnOnce<(rustc_middle[111bd832a8dd3e9b]::ty::context::TyCtxt, rustc_span[bc106c13571faf85]::def_id::LocalDefId)>>::call_once
  65:     0x7f6ac0cf4b6d - rustc_query_system[1b62b8d1dafbbc4d]::query::plumbing::try_execute_query::<rustc_query_impl[3b08f3ae92a78a66]::DynamicConfig<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::VecCache<rustc_span[bc106c13571faf85]::def_id::LocalDefId, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3b08f3ae92a78a66]::plumbing::QueryCtxt, false>
  66:     0x7f6ac2569bc1 - rustc_query_impl[3b08f3ae92a78a66]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  67:     0x7f6ac1fa5815 - rustc_hir_analysis[28d1c0fe6a280d9e]::collect::type_of::opaque::find_opaque_ty_constraints_for_rpit
  68:     0x7f6ac1fa2e45 - rustc_hir_analysis[28d1c0fe6a280d9e]::collect::type_of::type_of
  69:     0x7f6ac0d1e707 - rustc_query_impl[3b08f3ae92a78a66]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3b08f3ae92a78a66]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>
  70:     0x7f6ac0d1e440 - <rustc_query_impl[3b08f3ae92a78a66]::query_impl::type_of::dynamic_query::{closure#2} as core[7eaf5f0da9577636]::ops::function::FnOnce<(rustc_middle[111bd832a8dd3e9b]::ty::context::TyCtxt, rustc_span[bc106c13571faf85]::def_id::DefId)>>::call_once
  71:     0x7f6ac0c23ce5 - rustc_query_system[1b62b8d1dafbbc4d]::query::plumbing::try_execute_query::<rustc_query_impl[3b08f3ae92a78a66]::DynamicConfig<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::DefaultCache<rustc_span[bc106c13571faf85]::def_id::DefId, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3b08f3ae92a78a66]::plumbing::QueryCtxt, false>
  72:     0x7f6ac2566b23 - rustc_query_impl[3b08f3ae92a78a66]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  73:     0x7f6ac1a0b930 - rustc_middle[111bd832a8dd3e9b]::query::plumbing::query_get_at::<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::DefaultCache<rustc_span[bc106c13571faf85]::def_id::DefId, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 8usize]>>>
  74:     0x7f6ac1a05951 - rustc_hir_analysis[28d1c0fe6a280d9e]::check::check::check_mod_item_types
  75:     0x7f6ac18b934c - rustc_query_impl[3b08f3ae92a78a66]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3b08f3ae92a78a66]::query_impl::check_mod_item_types::dynamic_query::{closure#2}::{closure#0}, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 0usize]>>
  76:     0x7f6ac18b932e - <rustc_query_impl[3b08f3ae92a78a66]::query_impl::check_mod_item_types::dynamic_query::{closure#2} as core[7eaf5f0da9577636]::ops::function::FnOnce<(rustc_middle[111bd832a8dd3e9b]::ty::context::TyCtxt, rustc_span[bc106c13571faf85]::def_id::LocalDefId)>>::call_once
  77:     0x7f6ac13329bf - rustc_query_system[1b62b8d1dafbbc4d]::query::plumbing::try_execute_query::<rustc_query_impl[3b08f3ae92a78a66]::DynamicConfig<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::VecCache<rustc_span[bc106c13571faf85]::def_id::LocalDefId, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[3b08f3ae92a78a66]::plumbing::QueryCtxt, false>
  78:     0x7f6ac2571884 - rustc_query_impl[3b08f3ae92a78a66]::query_impl::check_mod_item_types::get_query_non_incr::__rust_end_short_backtrace
  79:     0x7f6ac2080417 - <rustc_middle[111bd832a8dd3e9b]::hir::map::Map>::for_each_module::<rustc_hir_analysis[28d1c0fe6a280d9e]::check_crate::{closure#6}::{closure#0}>
  80:     0x7f6ac207f84d - <rustc_session[6c3c6789374dfdb7]::session::Session>::time::<(), rustc_hir_analysis[28d1c0fe6a280d9e]::check_crate::{closure#6}>
  81:     0x7f6ac207db77 - rustc_hir_analysis[28d1c0fe6a280d9e]::check_crate
  82:     0x7f6ac2078aad - rustc_interface[ec6df654b897cc82]::passes::analysis
  83:     0x7f6ac210330a - rustc_query_impl[3b08f3ae92a78a66]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3b08f3ae92a78a66]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 1usize]>>
  84:     0x7f6ac21032f9 - <rustc_query_impl[3b08f3ae92a78a66]::query_impl::analysis::dynamic_query::{closure#2} as core[7eaf5f0da9577636]::ops::function::FnOnce<(rustc_middle[111bd832a8dd3e9b]::ty::context::TyCtxt, ())>>::call_once
  85:     0x7f6ac2283838 - rustc_query_system[1b62b8d1dafbbc4d]::query::plumbing::try_execute_query::<rustc_query_impl[3b08f3ae92a78a66]::DynamicConfig<rustc_query_system[1b62b8d1dafbbc4d]::query::caches::SingleCache<rustc_middle[111bd832a8dd3e9b]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[3b08f3ae92a78a66]::plumbing::QueryCtxt, false>
  86:     0x7f6ac2283609 - rustc_query_impl[3b08f3ae92a78a66]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  87:     0x7f6ac225b8f5 - <rustc_middle[111bd832a8dd3e9b]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[bc9e93faacc37431]::run_compiler::{closure#1}::{closure#2}::{closure#4}, core[7eaf5f0da9577636]::result::Result<(), rustc_span[bc106c13571faf85]::ErrorGuaranteed>>
  88:     0x7f6ac1eae6ef - <rustc_interface[ec6df654b897cc82]::interface::Compiler>::enter::<rustc_driver_impl[bc9e93faacc37431]::run_compiler::{closure#1}::{closure#2}, core[7eaf5f0da9577636]::result::Result<core[7eaf5f0da9577636]::option::Option<rustc_interface[ec6df654b897cc82]::queries::Linker>, rustc_span[bc106c13571faf85]::ErrorGuaranteed>>
  89:     0x7f6ac1eac289 - <scoped_tls[ada14302b38d1f15]::ScopedKey<rustc_span[bc106c13571faf85]::SessionGlobals>>::set::<rustc_interface[ec6df654b897cc82]::interface::run_compiler<core[7eaf5f0da9577636]::result::Result<(), rustc_span[bc106c13571faf85]::ErrorGuaranteed>, rustc_driver_impl[bc9e93faacc37431]::run_compiler::{closure#1}>::{closure#0}, core[7eaf5f0da9577636]::result::Result<(), rustc_span[bc106c13571faf85]::ErrorGuaranteed>>
  90:     0x7f6ac1eab6f6 - std[74cc9e3c59aa80ae]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[ec6df654b897cc82]::util::run_in_thread_pool_with_globals<rustc_interface[ec6df654b897cc82]::interface::run_compiler<core[7eaf5f0da9577636]::result::Result<(), rustc_span[bc106c13571faf85]::ErrorGuaranteed>, rustc_driver_impl[bc9e93faacc37431]::run_compiler::{closure#1}>::{closure#0}, core[7eaf5f0da9577636]::result::Result<(), rustc_span[bc106c13571faf85]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[7eaf5f0da9577636]::result::Result<(), rustc_span[bc106c13571faf85]::ErrorGuaranteed>>
  91:     0x7f6ac1eab4a5 - <<std[74cc9e3c59aa80ae]::thread::Builder>::spawn_unchecked_<rustc_interface[ec6df654b897cc82]::util::run_in_thread_pool_with_globals<rustc_interface[ec6df654b897cc82]::interface::run_compiler<core[7eaf5f0da9577636]::result::Result<(), rustc_span[bc106c13571faf85]::ErrorGuaranteed>, rustc_driver_impl[bc9e93faacc37431]::run_compiler::{closure#1}>::{closure#0}, core[7eaf5f0da9577636]::result::Result<(), rustc_span[bc106c13571faf85]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[7eaf5f0da9577636]::result::Result<(), rustc_span[bc106c13571faf85]::ErrorGuaranteed>>::{closure#1} as core[7eaf5f0da9577636]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  92:     0x7f6ac3da13a5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5622047e08c7cdf0
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/alloc/src/boxed.rs:1985:9
  93:     0x7f6ac3da13a5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5e7463f2383057fb
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/alloc/src/boxed.rs:1985:9
  94:     0x7f6ac3da13a5 - std::sys::unix::thread::Thread::new::thread_start::h5c3763b970acc3be
                               at /rustc/df77afbcaf3365a32066a8ca4a00ae6fc9a69647/library/std/src/sys/unix/thread.rs:108:17
  95:     0x7f6abf89344b - <unknown>
  96:     0x7f6abf916e40 - <unknown>
  97:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.72.0-nightly (df77afbca 2023-06-12) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [mir_built] building MIR for `something::{closure#0}::{closure#0}`
#1 [unsafety_check_result] unsafety-checking `something::{closure#0}::{closure#0}`
#2 [unsafety_check_result] unsafety-checking `something::{closure#0}`
#3 [unsafety_check_result] unsafety-checking `something`
#4 [mir_const] preparing `something` for borrow checking
#5 [mir_promoted] promoting constants in MIR for `something`
#6 [mir_borrowck] borrow-checking `something`
#7 [type_of] computing type of `something::{opaque#0}`
#8 [check_mod_item_types] checking item types in top-level module
#9 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 3 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0425`.
```

</p>
</details>

--- a/compiler/rustc_hir_typeck/src/expr_use_visitor.rs
+++ b/compiler/rustc_hir_typeck/src/expr_use_visitor.rs
@@ -443,7 +443,7 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
                         if matches!((lhs, wild, rhs), (&[], Some(_), &[]))
                             // Arrays have a statically known size, so
                             // there is no need to read their length
-                            || discr_place.place.base_ty.is_array()
+                            || place.place.ty().peel_refs().is_array()
                         {
                         } else {
                             needs_to_be_read = true;

--- a/tests/ui/closures/2229_closure_analysis/match/patterns-capture-analysis.rs
+++ b/tests/ui/closures/2229_closure_analysis/match/patterns-capture-analysis.rs
@@ -87,6 +87,31 @@ fn test_4_should_not_capture_array() {
         }
     };
     c();
+
+    // We also do not need to capture an array
+    // behind a reference (#112607)
+    let array: &[i32; 3] = &[0; 3];
+    let c = #[rustc_capture_analysis]
+    || {
+    //~^ First Pass analysis includes:
+        match array {
+            [_, _, _] => {}
+        }
+    };
+    c();
+
+    // We should still not insert a read if the array is inside an
+    // irrefutable pattern
+    struct Foo<T>(T);
+    let f = &Foo(&[10; 3]);
+    let c = #[rustc_capture_analysis]
+    || {
+    //~^ First Pass analysis includes:
+        match f {
+            Foo([_, _, _]) => ()
+        }
+    };
+    c();
 }
 
 // Testing MultiVariant patterns

--- a/tests/ui/closures/2229_closure_analysis/match/patterns-capture-analysis.stderr
+++ b/tests/ui/closures/2229_closure_analysis/match/patterns-capture-analysis.stderr
@@ -109,7 +109,29 @@ LL | |     };
    | |_____^
 
 error: First Pass analysis includes:
-  --> $DIR/patterns-capture-analysis.rs:105:5
+  --> $DIR/patterns-capture-analysis.rs:95:5
+   |
+LL | /     || {
+LL | |
+LL | |         match array {
+LL | |             [_, _, _] => {}
+LL | |         }
+LL | |     };
+   | |_____^
+
+error: First Pass analysis includes:
+  --> $DIR/patterns-capture-analysis.rs:108:5
+   |
+LL | /     || {
+LL | |
+LL | |         match f {
+LL | |             Foo([_, _, _]) => ()
+LL | |         }
+LL | |     };
+   | |_____^
+
+error: First Pass analysis includes:
+  --> $DIR/patterns-capture-analysis.rs:130:5
    |
 LL | /     || {
 LL | |
@@ -121,13 +143,13 @@ LL | |     };
    | |_____^
    |
 note: Capturing variant[] -> ImmBorrow
-  --> $DIR/patterns-capture-analysis.rs:108:15
+  --> $DIR/patterns-capture-analysis.rs:133:15
    |
 LL |         match variant {
    |               ^^^^^^^
 
 error: Min Capture analysis includes:
-  --> $DIR/patterns-capture-analysis.rs:105:5
+  --> $DIR/patterns-capture-analysis.rs:130:5
    |
 LL | /     || {
 LL | |
@@ -139,13 +161,13 @@ LL | |     };
    | |_____^
    |
 note: Min Capture variant[] -> ImmBorrow
-  --> $DIR/patterns-capture-analysis.rs:108:15
+  --> $DIR/patterns-capture-analysis.rs:133:15
    |
 LL |         match variant {
    |               ^^^^^^^
 
 error: First Pass analysis includes:
-  --> $DIR/patterns-capture-analysis.rs:123:5
+  --> $DIR/patterns-capture-analysis.rs:148:5
    |
 LL | /     || {
 LL | |
@@ -157,13 +179,13 @@ LL | |     };
    | |_____^
    |
 note: Capturing slice[] -> ImmBorrow
-  --> $DIR/patterns-capture-analysis.rs:126:15
+  --> $DIR/patterns-capture-analysis.rs:151:15
    |
 LL |         match slice {
    |               ^^^^^
 
 error: Min Capture analysis includes:
-  --> $DIR/patterns-capture-analysis.rs:123:5
+  --> $DIR/patterns-capture-analysis.rs:148:5
    |
 LL | /     || {
 LL | |
@@ -175,13 +197,13 @@ LL | |     };
    | |_____^
    |
 note: Min Capture slice[] -> ImmBorrow
-  --> $DIR/patterns-capture-analysis.rs:126:15
+  --> $DIR/patterns-capture-analysis.rs:151:15
    |
 LL |         match slice {
    |               ^^^^^
 
 error: First Pass analysis includes:
-  --> $DIR/patterns-capture-analysis.rs:135:5
+  --> $DIR/patterns-capture-analysis.rs:160:5
    |
 LL | /     || {
 LL | |
@@ -193,13 +215,13 @@ LL | |     };
    | |_____^
    |
 note: Capturing slice[] -> ImmBorrow
-  --> $DIR/patterns-capture-analysis.rs:138:15
+  --> $DIR/patterns-capture-analysis.rs:163:15
    |
 LL |         match slice {
    |               ^^^^^
 
 error: Min Capture analysis includes:
-  --> $DIR/patterns-capture-analysis.rs:135:5
+  --> $DIR/patterns-capture-analysis.rs:160:5
    |
 LL | /     || {
 LL | |
@@ -211,13 +233,13 @@ LL | |     };
    | |_____^
    |
 note: Min Capture slice[] -> ImmBorrow
-  --> $DIR/patterns-capture-analysis.rs:138:15
+  --> $DIR/patterns-capture-analysis.rs:163:15
    |
 LL |         match slice {
    |               ^^^^^
 
 error: First Pass analysis includes:
-  --> $DIR/patterns-capture-analysis.rs:147:5
+  --> $DIR/patterns-capture-analysis.rs:172:5
    |
 LL | /     || {
 LL | |
@@ -229,13 +251,13 @@ LL | |     };
    | |_____^
    |
 note: Capturing slice[] -> ImmBorrow
-  --> $DIR/patterns-capture-analysis.rs:150:15
+  --> $DIR/patterns-capture-analysis.rs:175:15
    |
 LL |         match slice {
    |               ^^^^^
 
 error: Min Capture analysis includes:
-  --> $DIR/patterns-capture-analysis.rs:147:5
+  --> $DIR/patterns-capture-analysis.rs:172:5
    |
 LL | /     || {
 LL | |
@@ -247,13 +269,13 @@ LL | |     };
    | |_____^
    |
 note: Min Capture slice[] -> ImmBorrow
-  --> $DIR/patterns-capture-analysis.rs:150:15
+  --> $DIR/patterns-capture-analysis.rs:175:15
    |
 LL |         match slice {
    |               ^^^^^
 
 error: First Pass analysis includes:
-  --> $DIR/patterns-capture-analysis.rs:164:5
+  --> $DIR/patterns-capture-analysis.rs:189:5
    |
 LL | /     || {
 LL | |
@@ -264,5 +286,5 @@ LL | |         }
 LL | |     };
    | |_____^
 
-error: aborting due to 16 previous errors
+error: aborting due to 18 previous errors
 

